### DISCUSSION TITLE
Setting Tab Title and lua global variables

### DISCUSSION THREAD

**Discussion by sboesebeck** [wezterm/wezterm#2983] [General]
I am trying to get the different workarounds mentioned [here](https://github.com/wez/wezterm/issues/1598#issuecomment-1379261296) running somehow - but how can I change a value in the GLOBAL-Table called `tab_titles` at runtime? 

it seems like, when doing it via the debug overlay, it won't be written?

This is the code in my wezterm.lua:

```lua
if wezterm.GLOBAL.tab_titles == nil then
    wezterm.GLOBAL.tab_titles={}
end
wezterm.on("format-tab-title", function(tab)
	local title = tab.tab_id .. ": " .. "something new"
	if wezterm.GLOBAL.tab_titles["T" .. tab.tab_id] ~= nil then
	   title = wezterm.GLOBAL.tab_titles["T" .. tab.tab_id]
	end
	wezterm.log_info("Title: " .. title)
	return title
end)
```
when opening the debug overlay, this happens:

```
> wezterm.GLOBAL.tab_titles["T0"]="Test"
> > wezterm.GLOBAL.tab_titles
{}
```

EDIT: the function `set_tab_title` from @wez did not work for me at all. This method at least changed the names
if there is a better solution to rename tabs, this is of course obsolete...

Also: can we call those methods from "extern" - like when opening a new Tab run that lua to change the tab name?

**Comment by wez** [comment]
What's happening here is that `GLOBAL.something` (or `GLOBAL["something"]`, which is equivalent), returns a copy of the `something` key when you refer to it, rather than a reference to the underlying data.

So when you do:

```lua
wezterm.GLOBAL.tab_titles["T0"]="Test"
```

it's conceptually equivalent to:

```lua
temp = copy(wezterm.GLOBAL.tab_titles)
temp["T0"] = "Test"
```

and doesn't update the global.

Using distinct read, modify and then write steps should work, like shown in https://github.com/wez/wezterm/issues/1598#issuecomment-1167761301:

```lua
local current_titles = wezterm.GLOBAL.tab_titles
current_titles[tab_id] = title
wezterm.GLOBAL.tab_titles = current_titles
```

It is possible to do some work to make that referencing work, but it's a bit tricky because `wezterm.GLOBAL` is a specialized map that exists outside of the lua runtime environment, and we'd need to add a special userdata class to represent that sort of table value.  It needs some thought.

**Reply by wez** [reply]
9d7e613cfa22846f65162dd3a3a277f7d69d270d makes the referencing work as you might expect

**Reply by sboesebeck** [reply]
awesome thanks!
